### PR TITLE
prevent Code Notes vandalism - closes #309

### LINKED ISSUE #309: Require Developer permission to update/delete existing code notes
Recently we had a vandalism attack where a disgruntled user was maliciously deleting code notes from games. Because we allow any user to create/update code notes (even unregistered ones), and deleting is just updating to blank, he was able to sabotage several sets before we started banning him and the slew of alts he created.

I know that there's a whole permission model being added in v2, but we need a stopgap solution for v1. I recommend adding logic to prevent modification of existing notes if the user's rank is less than Developer. If possible, the user should be allowed to modify/delete their own notes.

### PR DESCRIPTION
**Pull request by meleu**
When trying to submit a new code note, do **NOT** allow the note to be updated/submitted if:

user is not a developer
**AND**
user is not the author of original note
**AND**
code note is not empty

In other words:
regular users can only submit/edit Code Notes when they are the original Author or the Note is empty.

### CODE REVIEW

**Comment by meleu** [comment]
**Drawback**: well-intentioned jr-devs won't be able to edit old Code Notes.

**Workaround**: they'll must ask a Developer to clear the Note and then it'll be free for jr-dev edition.

**Comment by Keltron3030** [comment]
Perfect solution, thanks Meleu.

File: lib/database/utility.php [hunk 1]
```diff
@@ -624,6 +626,20 @@ function SubmitCodeNote2($user, $gameID, $address, $note)
         return false;
     }
 
+    $addressHex = '0x' . str_pad(dechex($address), 6, '0', STR_PAD_LEFT);
+    $currentNotes = getCodeNotesData($gameID);
+    $i = array_search($addressHex, array_column($currentNotes, 'Address'));
+
+    if (
+        $i === false
```
Comment (Jamiras) [inline lib/database/utility.php]: doesn't `array_search` return false if the item is _not_ found?

Comment (meleu) [reply lib/database/utility.php]: yep, and if item is not found that address is available for a regular user to edit.

Comment (Jamiras) [reply lib/database/utility.php]: seems like your logic is backwards. combined with the following lines, I'm reading:

"if not found && not a developer && not the current user && not empty"

shouldn't it be

"if found && not a developer && not the current user && not empty"?

Comment (meleu) [reply lib/database/utility.php]: oh sh!t, you're right! it should be `$i !== false`.

**Comment by meleu** [comment]
After last commit fixing the logic I think I tested every possible situation locally and everything seems to be working fine. I'm going to put it on production. Please let me know if I broke something (via issue tracker).

I think it deserves an announcement for the jr-devs.

**Pull request merged**